exporter/azuredataexplorerexporter/adx_exporter.go

```diff
@@ -0,0 +1,281 @@
+// Copyright The OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//       http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package azuredataexplorerexporter // import "github.com/open-telemetry/opentelemetry-collector-contrib/exporter/azuredataexplorerexporter"
+
+import (
+	"context"
+	"io"
+	"strings"
+
+	"github.com/Azure/azure-kusto-go/kusto"
+	"github.com/Azure/azure-kusto-go/kusto/data/errors"
+	"github.com/Azure/azure-kusto-go/kusto/ingest"
+	"github.com/Azure/go-autorest/autorest/azure/auth"
+	jsoniter "github.com/json-iterator/go"
+	"go.opentelemetry.io/collector/pdata/pcommon"
+	"go.opentelemetry.io/collector/pdata/plog"
+	"go.opentelemetry.io/collector/pdata/pmetric"
+	"go.opentelemetry.io/collector/pdata/ptrace"
+	"go.uber.org/zap"
+)
+
+// adxDataProducer uses the ADX client to perform ingestion
+type adxDataProducer struct {
+	client        *kusto.Client       // client for logs , traces and metrics
+	ingestor      localIngestor       // ingestion for  logs, traces and metrics
+	ingestoptions []ingest.FileOption // Options for the ingestion
+	logger        *zap.Logger         // Loggers for tracing the flow
+}
+
+type localIngestor interface {
+	FromReader(ctx context.Context, reader io.Reader, options ...ingest.FileOption) (*ingest.Result, error)
+	Close() error
+}
+
+const nextline = "\n"
+
+const (
+	// Scope name
+	scopename = "scope.name"
+	// Scope version
+	scopeversion = "scope.version"
+)
+
+// given the full metrics , extract each metric , resource attributes and scope attributes. Individual metric mapping is sent on to metricdata mapping
+func (e *adxDataProducer) metricsDataPusher(ctx context.Context, metrics pmetric.Metrics) error {
+	var metricsBuffer []string
+	transformedadxmetrics, err := rawMetricsToAdxMetrics(ctx, metrics, e.logger)
+	if err != nil {
+		e.logger.Error("Error transforming metrics to ADX metric format.", zap.Error(err))
+		return err
+	}
+	// Since the transform succeeded ,  using the option for ingestion ingest the data into ADX
+	for _, tm := range transformedadxmetrics {
+		adxmetricjsonstring, err := jsoniter.MarshalToString(tm)
+		if err != nil {
+			e.logger.Error("Error performing serialization of data.", zap.Error(err))
+		}
+		metricsBuffer = append(metricsBuffer, adxmetricjsonstring)
```

Comment:
Append will double the array size every time its capacity is reached.
Not a performance expert on Go, but do check use of Writer instead

Reply:
@ohadbitt Initial thought was same as you have mentioned, taking the data and writing it to Kusto for every metric element. Considering the points mentioned by @yogilad in the mail conversation with Quanmao LIU <qualiu@microsoft.com> that said, Kusto is not optimized for single line ingestion and not built to intake very small data items at high capacity, we moved to this strategy of holding up and collectively send the metric data to Kusto cluster.